src/DomainGuest.sol

```diff
@@ -0,0 +1,388 @@
+        emit File(what, data);
+    }
+
+    function file(bytes32 what, uint256 data) external auth {
+        if (what == "dust") dust = data;
+        else revert("DomainGuest/file-unrecognized-param");
+        emit File(what, data);
+    }
+
+    // --- Canonical DAI Support ---
+
+    /// @notice Mint DAI and send to user
+    /// @param to The address to send the DAI to on the local domain
+    /// @param amount The amount of DAI to send [WAD]
+    function deposit(address to, uint256 amount) external hostOnly {
+        vat.swell(address(this), _int256(amount * RAY));
+        daiJoin.exit(to, amount);
+
+        emit Deposit(to, amount);
+    }
+
+    /// @notice Withdraw DAI by burning local canonical DAI
+    /// @param to The address to send the DAI to on the remote domain
+    /// @param amount The amount of DAI to withdraw [WAD]
+    function _withdraw(address to, uint256 amount) internal returns (bytes memory payload) {
+        require(dai.transferFrom(msg.sender, address(this), amount), "DomainGuest/transfer-failed");
+        daiJoin.join(address(this), amount);
+        vat.swell(address(this), -_int256(amount * RAY));
+
+        payload = abi.encodeWithSelector(DomainHostLike.withdraw.selector, to, amount);
+
+        emit Withdraw(to, amount);
+    }
+
+    // --- MCD Support ---
+
+    /// @notice Record changes in line and grain and update dss global debt ceiling if necessary
+    /// @param _lid Local ordering id
+    /// @param wad The new debt ceiling [WAD]
+    function lift(uint256 _lid, uint256 wad) external hostOnly ordered(_lid) {
+        require(live == 1, "DomainGuest/not-live");
+
+        uint256 lastLine = vat.Line();
+        uint256 rad = wad * RAY;
+        if (rad > lastLine) grain = wad;
+        vat.file("Line", rad);
+
+        emit Lift(wad);
+    }
+
+    /// @notice Will release remote DAI from the escrow when it is safe to do so
+    /// @dev    Should be run by keeper on a regular schedule.
+    function _release() internal returns (bytes memory payload) {
+        require(live == 1, "DomainGuest/not-live");
+
+        uint256 limit = _max(vat.Line() / RAY, _divup(vat.debt(), RAY));
+        require(grain >= limit + dust / RAY, "DomainGuest/dust");
+        uint256 burned = grain - limit;
+        grain = limit;
+
+        payload = abi.encodeWithSelector(DomainHostLike.release.selector, rid++, burned);
+
+        emit Release(burned);
+    }
+
+    /// @notice Push surplus (or deficit) to the host dss
+    /// @dev Should be run by keeper on a regular schedule
+    function _push() internal returns (bytes memory payload) {
+        require(live == 1, "DomainGuest/not-live");
+
+        uint256 _dai = vat.dai(address(this));
+        uint256 _sin = vat.sin(address(this));
+        if (_dai >= _sin + dust) {
+            // We have a surplus
+            if (_sin > 0) vat.heal(_sin);
+
+            uint256 wad = (_dai - _sin) / RAY;    // Leave the dust
```

Comment:
Outdated comment? The dust is being pushed as well (which is correct imo).

Reply:
Yeah I got confused about it as well, but it means the "dust" of diving by `RAY` (rounds down).

Reply:
Ah, makes sense :)

Reply:
Ah yes the actual `dust` variable was added after that comment so as Gonza says the comment is confusing now. I'll update it.

Reply:
Updated to something a bit more clear.